p2p/protocol/autonatv2/server.go

```diff
@@ -0,0 +1,430 @@
+		s.Reset()
+		log.Debugf("invalid message type from %s: %T expected: DialRequest", p, msg.Msg)
+		return
+	}
+
+	// parse peer's addresses
+	var dialAddr ma.Multiaddr
+	var addrIdx int
+	for i, ab := range msg.GetDialRequest().GetAddrs() {
+		if i >= maxPeerAddresses {
+			break
+		}
+		a, err := ma.NewMultiaddrBytes(ab)
+		if err != nil {
+			continue
+		}
+		if !as.allowPrivateAddrs && !manet.IsPublicAddr(a) {
+			continue
+		}
+		if !as.dialerHost.Network().CanDial(p, a) {
+			continue
+		}
+		dialAddr = a
+		addrIdx = i
+		break
+	}
+	// No dialable address
+	if dialAddr == nil {
+		msg = pb.Message{
+			Msg: &pb.Message_DialResponse{
+				DialResponse: &pb.DialResponse{
+					Status: pb.DialResponse_E_DIAL_REFUSED,
+				},
+			},
+		}
+		if err := w.WriteMsg(&msg); err != nil {
+			s.Reset()
+			log.Debugf("failed to write dial refused response to %s: %s", p, err)
+			return
+		}
+		return
+	}
+
+	nonce := msg.GetDialRequest().Nonce
+
+	isDialDataRequired := as.dialDataRequestPolicy(s, dialAddr)
+	if isDialDataRequired && !as.limiter.AcceptDialDataRequest(p) {
+		msg = pb.Message{
+			Msg: &pb.Message_DialResponse{
+				DialResponse: &pb.DialResponse{
+					Status: pb.DialResponse_E_REQUEST_REJECTED,
+				},
+			},
+		}
+		if err := w.WriteMsg(&msg); err != nil {
+			s.Reset()
+			log.Debugf("failed to write request rejected response to %s: %s", p, err)
+			return
+		}
+		log.Debugf("rejected request from %s: rate limit exceeded", p)
+		return
+	}
+
+	if isDialDataRequired {
+		if err := getDialData(w, s, &msg, addrIdx); err != nil {
+			s.Reset()
+			log.Debugf("%s refused dial data request: %s", p, err)
+			return
+		}
+	}
+
+	dialStatus := as.dialBack(s.Conn().RemotePeer(), dialAddr, nonce)
+	msg = pb.Message{
+		Msg: &pb.Message_DialResponse{
+			DialResponse: &pb.DialResponse{
+				Status:     pb.DialResponse_OK,
+				DialStatus: dialStatus,
+				AddrIdx:    uint32(addrIdx),
+			},
+		},
+	}
+	if err := w.WriteMsg(&msg); err != nil {
+		s.Reset()
+		log.Debugf("failed to write response to %s: %s", p, err)
+		return
+	}
+}
+
+// getDialData gets data from the client for dialing the address
+func getDialData(w pbio.Writer, s network.Stream, msg *pb.Message, addrIdx int) error {
+	numBytes := minHandshakeSizeBytes + rand.Intn(maxHandshakeSizeBytes-minHandshakeSizeBytes)
+	*msg = pb.Message{
+		Msg: &pb.Message_DialDataRequest{
+			DialDataRequest: &pb.DialDataRequest{
+				AddrIdx:  uint32(addrIdx),
+				NumBytes: uint64(numBytes),
+			},
+		},
+	}
+	if err := w.WriteMsg(msg); err != nil {
+		return fmt.Errorf("dial data write: %w", err)
+	}
+	// pbio.Reader that we used so far on this stream is buffered. But at this point
+	// there is nothing unread on the stream. So it is safe to use the raw stream to
+	// read, reducing allocations.
+	return readDialData(numBytes, s)
+}
+
+func readDialData(numBytes int, r io.Reader) error {
+	mr := &msgReader{R: r, Buf: make([]byte, maxMsgSize)}
+	for remain := numBytes; remain > 0; {
```

Comment:
Using 	`pool "github.com/libp2p/go-buffer-pool"`. Saves you one alloc.
```suggestion
	buf := pool.Get(maxMsgSize)
	defer pool.Put(buf)
	mr := &msgReader{R: r, Buf: buf}
```